### ISSUE TITLE
Poor performance of the model when enabling layer normalization with tf.contrib.rnn.LayerNormBasicLSTMCell

### ISSUE THREAD

**Issue by AzizCode92** [tensorflow/tensorflow#21929]
-------------------------------------------------------------------------------------------------------

### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: No

- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: 

VERSION="16.04.4 LTS (Xenial Xerus)"
VERSION_ID="16.04"
VERSION_CODENAME=xenial

- **Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device**: 
No
- **TensorFlow installed from (source or binary)**:
 source
- **TensorFlow version (use command below)**:  
tf.VERSION = 1.9.0

- **Python version**: python 2.7

- **Bazel version (if compiling from source)**:  
0.11.1

- **GCC/Compiler version (if compiling from source)**:  
c++ (Ubuntu 5.4.0-6ubuntu1~16.04.10) 5.4.0 20160609

- **CUDA/cuDNN version**: 
 9.1
- **GPU model and memory**:  
Tesla K80
- **Exact command to reproduce**:
( No command)

-------------------------------------------------------------------------------------------------------
Before writing my issue here,  I found a similar question on  [stackoverflow]( https://stackoverflow.com/questions/45150101/why-is-layernormbasiclstmcell-much-slower-and-less-accurate-than-lstmcell) where the author described the same problem I have but there were no valid answer and most of the comments where wrongly referring to technique related to batch_normalization which are non-applicable since we are talking about layer_normalization. 
( [Difference between layer_normalzation and batch_normalization](https://theneuralperspective.com/2016/10/27/gradient-topics/) )
I was trying to do some layer normalization with my model, and I have chosen the 
`tf.contrib.rnn.LayerNormBasicLSTMCell` where `layer_norm` is set to True.
I started training the model and as unexpected the training/validation losses are decreasing slowly.
The paper of layer normalization ( https://arxiv.org/pdf/1607.06450.pdf ) showed that this technique makes the training time much faster but the Tensorflow implementation showed the inverse case.
Are there please any hints on how to solve this problem? Any further explications on why the implementation makes the training slower?

**Comment by AzizCode92** [comment]
Here I added some results of the experiment I have done using the MNIST dataset.

> def RNN(x, weights, biases):

    # Prepare data shape to match `rnn` function requirements
    # Current data input shape: (batch_size, timesteps, n_input)
    # Required shape: 'timesteps' tensors list of shape (batch_size, n_input)

    # Unstack to get a list of 'timesteps' tensors of shape (batch_size, n_input)
    x = tf.unstack(x,timesteps,1)
    

    # Define a lstm cell with tensorflow
    lstm_cell = rnn.LayerNormBasicLSTMCell(num_hidden, forget_bias=1.0)
	# Get lstm cell output
    outputs, states = rnn.static_rnn(lstm_cell, x, dtype=tf.float32)

    # Linear activation, using rnn inner loop last output
    return tf.matmul(outputs[-1], weights['out']) + biases['out']

The output of the model was the following.   

> Step 1, Minibatch Loss= nan, Training Accuracy= 0.070
Step 200, Minibatch Loss= nan, Training Accuracy= 0.117
Step 400, Minibatch Loss= nan, Training Accuracy= 0.195
Step 600, Minibatch Loss= nan, Training Accuracy= 0.117
Step 800, Minibatch Loss= nan, Training Accuracy= 0.086
Step 1000, Minibatch Loss= nan, Training Accuracy= 0.102
Step 1200, Minibatch Loss= nan, Training Accuracy= 0.109
Step 1400, Minibatch Loss= nan, Training Accuracy= 0.148
Step 1600, Minibatch Loss= nan, Training Accuracy= 0.133
Step 1800, Minibatch Loss= nan, Training Accuracy= 0.102
Step 2000, Minibatch Loss= nan, Training Accuracy= 0.086
Step 2200, Minibatch Loss= nan, Training Accuracy= 0.094
Step 2400, Minibatch Loss= nan, Training Accuracy= 0.062
Step 2600, Minibatch Loss= nan, Training Accuracy= 0.094
Step 2800, Minibatch Loss= nan, Training Accuracy= 0.094
Step 3000, Minibatch Loss= nan, Training Accuracy= 0.109
Step 3200, Minibatch Loss= nan, Training Accuracy= 0.086
Step 3400, Minibatch Loss= nan, Training Accuracy= 0.109
Step 3600, Minibatch Loss= nan, Training Accuracy= 0.148
Step 3800, Minibatch Loss= nan, Training Accuracy= 0.102
Step 4000, Minibatch Loss= nan, Training Accuracy= 0.117
Step 4200, Minibatch Loss= nan, Training Accuracy= 0.078
Step 4400, Minibatch Loss= nan, Training Accuracy= 0.102
Step 4600, Minibatch Loss= nan, Training Accuracy= 0.062
Step 4800, Minibatch Loss= nan, Training Accuracy= 0.078
Step 5000, Minibatch Loss= nan, Training Accuracy= 0.070
Step 5200, Minibatch Loss= nan, Training Accuracy= 0.062
Step 5400, Minibatch Loss= nan, Training Accuracy= 0.125
Step 5600, Minibatch Loss= nan, Training Accuracy= 0.109
Step 5800, Minibatch Loss= nan, Training Accuracy= 0.125
Step 6000, Minibatch Loss= nan, Training Accuracy= 0.062
Step 6200, Minibatch Loss= nan, Training Accuracy= 0.102
Step 6400, Minibatch Loss= nan, Training Accuracy= 0.094
Step 6600, Minibatch Loss= nan, Training Accuracy= 0.094
Step 6800, Minibatch Loss= nan, Training Accuracy= 0.102
Step 7000, Minibatch Loss= nan, Training Accuracy= 0.078
Step 7200, Minibatch Loss= nan, Training Accuracy= 0.148
Step 7400, Minibatch Loss= nan, Training Accuracy= 0.117
Step 7600, Minibatch Loss= nan, Training Accuracy= 0.125
Step 7800, Minibatch Loss= nan, Training Accuracy= 0.117
Step 8000, Minibatch Loss= nan, Training Accuracy= 0.102
Step 8200, Minibatch Loss= nan, Training Accuracy= 0.125
Step 8400, Minibatch Loss= nan, Training Accuracy= 0.062
Step 8600, Minibatch Loss= nan, Training Accuracy= 0.055
Step 8800, Minibatch Loss= nan, Training Accuracy= 0.094
Step 9000, Minibatch Loss= nan, Training Accuracy= 0.172
Step 9200, Minibatch Loss= nan, Training Accuracy= 0.078
Step 9400, Minibatch Loss= nan, Training Accuracy= 0.148
Step 9600, Minibatch Loss= nan, Training Accuracy= 0.086
Step 9800, Minibatch Loss= nan, Training Accuracy= 0.094
Step 10000, Minibatch Loss= nan, Training Accuracy= 0.078
Optimization Finished!
elapsed time is 6732.706034
Testing Accuracy: 0.078125

**Comment by AzizCode92** [comment]
@asimshankar any updates about this issue?

**Comment by AzizCode92** [comment]
here is the full code 
```python
""" Recurrent Neural Network.

A Recurrent Neural Network (LSTM) implementation example using TensorFlow library.
This example is using the MNIST database of handwritten digits (http://yann.lecun.com/exdb/mnist/)

Links:
    [Long Short Term Memory](http://deeplearning.cs.cmu.edu/pdfs/Hochreiter97_lstm.pdf)
    [MNIST Dataset](http://yann.lecun.com/exdb/mnist/).

Author: Aymeric Damien
Project: https://github.com/aymericdamien/TensorFlow-Examples/
"""

from __future__ import print_function

import tensorflow as tf
from tensorflow.contrib import rnn

# Import MNIST data
from tensorflow.examples.tutorials.mnist import input_data
mnist = input_data.read_data_sets("/tmp/data/", one_hot=True)

'''
To classify images using a recurrent neural network, we consider every image
row as a sequence of pixels. Because MNIST image shape is 28*28px, we will then
handle 28 sequences of 28 steps for every sample.
'''

# Training Parameters
learning_rate = 0.001
training_steps = 10000
batch_size = 128
display_step = 200

# Network Parameters
num_input = 28 # MNIST data input (img shape: 28*28)
timesteps = 28 # timesteps
num_hidden = 128 # hidden layer num of features
num_classes = 10 # MNIST total classes (0-9 digits)

# tf Graph input
X = tf.placeholder("float", [None, timesteps, num_input])
Y = tf.placeholder("float", [None, num_classes])

# Define weights
weights = {
    'out': tf.Variable(tf.random_normal([num_hidden, num_classes]))
}
biases = {
    'out': tf.Variable(tf.random_normal([num_classes]))
}


def RNN(x, weights, biases):

    # Prepare data shape to match `rnn` function requirements
    # Current data input shape: (batch_size, timesteps, n_input)
    # Required shape: 'timesteps' tensors list of shape (batch_size, n_input)

    # Unstack to get a list of 'timesteps' tensors of shape (batch_size, n_input)
    x = tf.unstack(x, timesteps, 1)

    # Define a lstm cell with tensorflow
    lstm_cell = rnn.LayerNormBasicLSTMCell(num_hidden, forget_bias=1.0)

    # Get lstm cell output
    outputs, states = rnn.static_rnn(lstm_cell, x, dtype=tf.float32)

    # Linear activation, using rnn inner loop last output
    return tf.matmul(outputs[-1], weights['out']) + biases['out']

logits = RNN(X, weights, biases)
prediction = tf.nn.softmax(logits)

# Define loss and optimizer
loss_op = tf.reduce_mean(tf.nn.softmax_cross_entropy_with_logits(
    logits=logits, labels=Y))
optimizer = tf.train.GradientDescentOptimizer(learning_rate=learning_rate)
train_op = optimizer.minimize(loss_op)

# Evaluate model (with test logits, for dropout to be disabled)
correct_pred = tf.equal(tf.argmax(prediction, 1), tf.argmax(Y, 1))
accuracy = tf.reduce_mean(tf.cast(correct_pred, tf.float32))

# Initialize the variables (i.e. assign their default value)
init = tf.global_variables_initializer()

# Start training
with tf.Session() as sess:

    # Run the initializer
    sess.run(init)

    for step in range(1, training_steps+1):
        batch_x, batch_y = mnist.train.next_batch(batch_size)
        # Reshape data to get 28 seq of 28 elements
        batch_x = batch_x.reshape((batch_size, timesteps, num_input))
        # Run optimization op (backprop)
        sess.run(train_op, feed_dict={X: batch_x, Y: batch_y})
        if step % display_step == 0 or step == 1:
            # Calculate batch loss and accuracy
            loss, acc = sess.run([loss_op, accuracy], feed_dict={X: batch_x,
                                                                 Y: batch_y})
            print("Step " + str(step) + ", Minibatch Loss= " + \
                  "{:.4f}".format(loss) + ", Training Accuracy= " + \
                  "{:.3f}".format(acc))

    print("Optimization Finished!")

    # Calculate accuracy for 128 mnist test images
    test_len = 128
    test_data = mnist.test.images[:test_len].reshape((-1, timesteps, num_input))
    test_label = mnist.test.labels[:test_len]
    print("Testing Accuracy:", \
        sess.run(accuracy, feed_dict={X: test_data, Y: test_label}))```

**Comment by asimshankar** [comment]
@petrux @ebrevdo : Any advice here?

**Comment by AzizCode92** [comment]
@petrux , thank you for your answer. I have also done the layer normalization using the tf.contrib.rnn.LayerNormBasicLSTMCell in a speech-recogntition task and the model was slow, converges so early and the accuracy decreases by the end.

**Comment by AzizCode92** [comment]
Here is [the encoder implementation ](https://github.com/vrenkens/nabu/blob/master/nabu/neuralnetworks/models/ed_encoders/listener.py)
in [this line ](https://github.com/vrenkens/nabu/blob/19404f2453f43721d20ee234aeefb77801614208/nabu/neuralnetworks/models/ed_encoders/listener.py#L50) I have used a pblstm layer where the `tf.contrib.rnn.LayerNormBasicLSTMCell` was used with `layer_norm = True`.

**Comment by petrux** [comment]
I think that being slower is somehow intended, since you have more parameters to optimize and more calculation to perform. The fact that it converges faster is intended as well. At that point, you could lower you learning rate to avoid the loss in accuracy (always, just on top of my head).

**Comment by AzizCode92** [comment]
Thank you for pointing that out. Actually, I'm using a validation to tune the learning rate. If the next two validation losses are worse than the actual loss, the training rate will be divided by two.
So, even taking this in account the accuracy is still way worse than the one I have without layer_normlaization.

**Comment by AzizCode92** [comment]
Another point I'm thinking of is to not use the `tf.contrib.rnn.LayerNormBasicLSTMCell` in my layers but instead use `tf.contrib.layers.layer_norm` in the layers of my encoder/decoder. Would that make a sense?

**Comment by ebrevdo** [comment]
See also:

https://github.com/tensorflow/tensorflow/pull/15861

**Comment by AzizCode92** [comment]
Yes, I have considered his pull request and I got the same performance as the one we have.

**Comment by ebrevdo** [comment]
Looks like we're getting into the black magic of nn tweaking. It's not
clear that there's a bug in the layer itself but rather likely that
tweaking hyperparams or using layer norm elsewhere is the right approach.
If you set the learning rrate to 0 or very small, do you still get
instabilities?  If not, then it's unlikely to be caused by the layer.

**Comment by AzizCode92** [comment]
I have set the learning rate to `0.000001` and the results were the same with the `tf.contrib.rnn.LayerNormBasicLSTMCell`. 
The model terminates with an accuracy of `0.078125`.
The idea of using the layer norm elsewhere seems to be working pretty good in the case of the speech-recognition task ( the model is much faster and the accuracy are good, still training!!!)

**Comment by petrux** [comment]
Probably because speech recognition is an actual _recurrent_ problem. Maybe you might want to train a FF network on MNIST data and try to layer normalize?---like in the original paper.

**Comment by AzizCode92** [comment]
I will try it and let you updated.
thank you

**Comment by AzizCode92** [comment]
This is a simple neural network with 5 layers on the MNIST dataset.
At each layer I applied layer normalization using `tf.contrib.layers.layer_norm`

```python
from tensorflow.examples.tutorials.mnist import input_data
import tensorflow as tf

batch_size = 100
learning_rate = 0.003
training_epoch = 10

mnist = input_data.read_data_sets('data', one_hot=True)

X_ = tf.placeholder(tf.float32, [None, 784])
Y_ = tf.placeholder(tf.float32, [None, 10])

L = 200
M = 100
N = 60
O = 30

X = tf.reshape(X_, [-1, 784])
W1 = tf.Variable(tf.truncated_normal([784, L], stddev=0.1))
B1 = tf.Variable(tf.ones([L])/10)
W2 = tf.Variable(tf.truncated_normal([L, M], stddev=0.1))
B2 = tf.Variable(tf.ones([M])/10)
W3 = tf.Variable(tf.truncated_normal([M, N], stddev=0.1))
B3 = tf.Variable(tf.ones([N])/10)
W4 = tf.Variable(tf.truncated_normal([N, O], stddev=0.1))
B4 = tf.Variable(tf.ones(O)/10)
W5 = tf.Variable(tf.truncated_normal([O, 10], stddev=0.1))
B5 = tf.Variable(tf.zeros(10))

Y1 = tf.nn.relu(tf.matmul(X, W1) + B1)
Y1 = tf.contrib.layers.layer_norm(Y1)
Y2 = tf.nn.relu(tf.matmul(Y1, W2) + B2)
Y2 = tf.contrib.layers.layer_norm(Y2)
Y3 = tf.nn.relu(tf.matmul(Y2, W3) + B3)
Y3 = tf.contrib.layers.layer_norm(Y3)
Y4 = tf.nn.relu(tf.matmul(Y3, W4) + B4)
Y4 = tf.contrib.layers.layer_norm(Y4)
Ylogits = tf.matmul(Y4, W5) + B5
Y = tf.nn.softmax(Ylogits)


cross_entropy = tf.reduce_mean(tf.nn.softmax_cross_entropy_with_logits(labels=Y_, logits=Ylogits))
correct_prediction = tf.equal(tf.argmax(Y, 1), tf.argmax(Y_, 1))
accuracy = tf.reduce_mean(tf.cast(correct_prediction, tf.float32))
train_step = tf.train.AdamOptimizer(learning_rate).minimize(cross_entropy)

with tf.Session() as sess:
	sess.run(tf.global_variables_initializer())
	for epoch in range(training_epoch):
		batch_count = int(mnist.train.num_examples/batch_size)
		for i in range(batch_count):
			batch_x, batch_y = mnist.train.next_batch(batch_size)
			sess.run([train_step], feed_dict= {X: batch_x, Y_: batch_y})
		print("Epoch: ", epoch, ", Accuracy: ", accuracy.eval(feed_dict={X: mnist.test.images, Y_: mnist.test.labels}))

```
and the results were like this 

```
Epoch:  0 , Accuracy:  0.9589
Epoch:  1 , Accuracy:  0.9631
Epoch:  2 , Accuracy:  0.9721
Epoch:  3 , Accuracy:  0.9659
Epoch:  4 , Accuracy:  0.9718
Epoch:  5 , Accuracy:  0.9729
Epoch:  6 , Accuracy:  0.9729
Epoch:  7 , Accuracy:  0.9779
Epoch:  8 , Accuracy:  0.9796
Epoch:  9 , Accuracy:  0.9763
```

**Comment by petrux** [comment]
Indeed. Please note that you were tackling the same problem with two radical different solutions. thanks for sharing the result of this latest experiment.

**Comment by ebrevdo** [comment]
The tensor2tensor team has had success with layer norm in sequential models
but they use it in a very different way.  It probably depends on the
problem domain to see where it will work.

**Comment by AzizCode92** [comment]
Well, I tuned my code and now with layer normalization i have the following output  

> Step 1, Minibatch Loss= 4.9653, Training Accuracy= 0.086
Step 200, Minibatch Loss= 0.4167, Training Accuracy= 0.859
Step 400, Minibatch Loss= 0.1636, Training Accuracy= 0.969
I used the AdamOptimizer gradient

**Comment by yif0** [comment]
> Well, I tuned my code and now with layer normalization i have the following output
> 
> > Step 1, Minibatch Loss= 4.9653, Training Accuracy= 0.086
> > Step 200, Minibatch Loss= 0.4167, Training Accuracy= 0.859
> > Step 400, Minibatch Loss= 0.1636, Training Accuracy= 0.969
> 
> I used instead the AdamOptimizer gradient and of course a whole different implementation of layer normalization

Hi, if you don't mind can you point to which implementation of layer norm you used? Thanks!

**Comment by KyotoSunshine** [comment]
I had the same issue with tf.contrib.layers.layer_norm, getting NaNs during training. Here is how I solved it: change the variance epsilon from 1e-12 to 1e-3. There seems to be some numerical instability during gradient computation when there are many layer norms in a row and the input has zero variance. In that case there will be direct division by epsilon. Unfortunately currently variance epsilon is not exposed as an argument to the tf.contrib.layer_norm, however I made a pull request to expose it recently: https://github.com/tensorflow/tensorflow/pull/25813
